fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1850,39 +1909,42 @@ public static OnlineIndexer forRecordStoreAndIndex(@Nonnull FDBRecordStore recor
          */
-        BUILD_IF_DISABLED(false),
+        BUILD_IF_DISABLED,
         /**
-         * Build if the index is disabled; Continue build if the index is write-only.
+         * Build if the index is disabled; Continue build if the index is write-only - using the previous
+         * method that was used.
          * <p>
          * Recommended. This should be sufficient if current index data is not corrupted.
          * </p>
          */
-        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY(true),
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY,
         /**
          * Build if the index is disabled; Rebuild if the index is write-only.
          */
-        BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY(false),
+        BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY,
         /**
          * Rebuild the index anyway, no matter it it disabled or write-only or readable.
          */
-        FORCE_BUILD(false),
+        FORCE_BUILD,
         /**
-         * Error if the index is disabled, or continue to build if the index is write only. To use this option to build
-         * an index, one should mark the index as write-only and clear existing index entries before building. This
-         * option is provided to make {@link #buildIndexAsync()} (or its variations) behave same as what it did before
-         * version 2.8.90.0, which is not recommended. {@link #BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY} should be adopted instead.
+         * Continue to build only if the index is write only and the requested method matches the previous one.
+         * To use this option to build an index, one should mark the index as write-only and clear the existing
+         * index entries before building. This option is provided to make {@link #buildIndexAsync()} (or its
+         * variations) behave same as what it did before version 2.8.90.0, which is not recommended.
+         * {@link #BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY} should be adopted instead.
          */
-        ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY(true),
-        ;
-
-        private boolean continueIfWriteOnly;
-
-        IndexStatePrecondition(boolean continueIfWriteOnly) {
-            this.continueIfWriteOnly = continueIfWriteOnly;
-        }
+        ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY,
 
-        public boolean isContinueIfWriteOnly() {
-            return continueIfWriteOnly;
-        }
+        /**
+         * Build if the index is disabled; Continue build if the index is write-only - only if the requested
+         * method is the same as the previous one.
+         */
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_ERROR_IF_POLICY_CHANGED,
+        /**
+         * Build if the index is disabled; Continue build if the index is write-only - only if the requested
+         * method matches the previous one, else restart the built.
+         */
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_REBUILD_IF_POLICY_CHANGED,
+        ;
```

Comment:
It's hard to reason about all options, it's hard to make sure the code considers all options and (particularly) it's hard test all options. Considering part of the options are orthogonal, I'm thinking that it might be more clear to refactor the precondition to something like:
```java
    enum HandleIfDisabled {
        IGNORE,
        ERROR,
        REBUILD
    }
    enum HandleIfWriteOnly {
        IGNORE,
        ERROR,
        REBUILD,
        CONTINUE // Need ContinuationPolicy
    }
    enum HandleIfCannotContinue {
        IGNORE,
        ERROR,
        REBUILD
    }
    class IndexStatePrecondition {
        @Nonnull HandleIfDisabled handleIfDisabled;
        @Nonnull HandleIfWriteOnly handleIfWriteOnly;
        @Nullable HandleIfCannotContinue handleIfCannotContinue; // Use builder to make sure this value is set when AND only when handleIfWriteOnly == HandleIfCannotContinue
    }
```

Reply:
For the record, we agreed to leave it as is for now but to add tests for FORCE_BUILD and BUILD_IF_DISABLED.